### ISSUE TITLE
Header Block component

### ISSUE THREAD

**Issue by bhoppers2008** [jpmorganchase/salt-ds#3839]
This issue will standardize a Header Block which can be composed into container components (Dialog, Drawer, Overlay) and provide a Header Block component.

**Requirements**

The Header Block will include the following attributes:

- Default: Accent bar, Title text, Close Button (all optional)
- Supports statuses (iRAG)
-- Header Block won't be opinionated about which components can display its status states. Salt will provide relevant examples of use per component it is used in.
- Padding-top/ -bottom at `--spacing-300`.

**Component updates**

The following components are updated to use the Header Block:

- Dialog
-- This will be the only example that shows status headers.
- Drawer
- Overlay
-- Padding-top/-bottom will be overwritten to `--spacing-200` to match the internal padding of Overlay.

In all cases, if the container element needs to be closed via Close Button, the Header Block close button should be used.

_Consideration: Should a Close Button component be created as it's own individual component?_

**Enhancement**
An overflow drop-shadow will be added at a later point.

Closes #3532
Closes #2970

**Comment by origami-z** [comment]
Need to think through how to play nicely with container's padding. e.g. Dialog has only top/bottom padding to make accent bar work. Drawer & Overlay have full paddings for the content (different spacing as well).

**Comment by origami-z** [comment]
Extend to 2 sprints, given multiple components, and user validation needed

**Comment by tomhazledine** [comment]
Figma spec complete. Ready for dev work to begin.

**Comment by mark-tate** [comment]
Cortado Goal: RC by EOS

**Comment by dplsek** [comment]
@tomhazledine link to Figma [spec](https://www.figma.com/design/8msOmbLvIsTtaIcNcbAxiW/Header-block?node-id=2038-1016&t=xNOcioTQrT2WojUa-1)

**Comment by origami-z** [comment]
Aug 12 - starting dev tomorrow (one minor design update remaining)

**Comment by dplsek** [comment]
Aug 13th - had header block dev kick-off with @tomhazledine @navkaur76 @pseys 

Open questions:
1. if header block were used in overlay, how is it expected to work compared to how it currently does? (close button only vs requiring a header)
2. will header block also be used in status dialogs?
3. need to settle on a concept/name that works within the system for the different options (currently 200 and 300)

@bhoppers2008 @joshwooding would like to discuss more on these

**Comment by navkaur76** [comment]
Meeting w/ @bhoppers2008 on Aug 14

**Answers to Qs in Darrin's comment above:** 

1. Content within an overlay will sit below the header and close button, spanning the entire width of the content area. Header block component will have an additional prop that hides all text elements within the parent container, so only the close button is on display.
2. Yes - header block to be used in status dialogs. Nav to update component set to include status variants
3. Naming convention for spacing-200 + 300 still TBD (cc @pseys @dplsek). Further discussion required on this. 

Other:
- Plan required for best way of handling deprecation in figma of heading in dialog, without causing significant breaking changes. Nav to explore options.

**Comment by tomhazledine** [comment]
Aug 16 - dev in-progress. Expect RC by end of Monday

**Comment by joshwooding** [comment]
+1 sprint due to OOO and further design requirements

**Comment by joshwooding** [comment]
Aug 20 - Proposal created, ready for review by @bhoppers2008 and @joshwooding

**Comment by origami-z** [comment]
Feedback from MTK - existing Dialog header, if close button is positioned before the title, close button will not receive click event

```
<DialogCloseButton onClick={handleRequestClose} />
<DialogTitle header="Export" />
```

**Comment by mark-tate** [comment]
Espresso Goal to be released on 13th Sept, with a design review on 9th September

Deprecation strategy for design will be defined as part of this release

**Comment by tomhazledine** [comment]
Handover for design review by EOD 9th.

**Comment by mark-tate** [comment]
Nav to create a meeting to discuss the options for release by 13th Sept

**Comment by tomhazledine** [comment]
Following code review, implementing changes to component API to avoid "magic" and favour composition.

Branch for docs has been opened, but so far site docs and examples are WIP: https://github.com/jpmorganchase/salt-ds/pull/4095.

**Comment by mark-tate** [comment]
Frappe Goal: merged to Labs by EOS
- @navkaur76 to write site docs and design review
- how designs docs will be released

**Comment by tomhazledine** [comment]
Refactoring following dev feedback, expect PR ready for dev-review tomorrow (18th)

**Comment by mark-tate** [comment]
Design review today, PR up for code review, examples may need updates ?

**Comment by mark-tate** [comment]
design review complete for dialog and overlay, outstanding issue with scrolling 
(long content with Safari chrome has no keyboard support)
Take a defect on scrolling to avoid Header blocker

**Comment by tomhazledine** [comment]
Code changes following design review in-progress

**Comment by mark-tate** [comment]
@tomhazledine  do we have an update on this ?

**Comment by mark-tate** [comment]
Latte: 1 week capacity
wk1: Header for Overlay & Dialog available in Labs
To move to core some work needed 
@bhoppers2008 to schedule remaining work

**Comment by bhoppers2008** [comment]
No work to happen beyond pushing dialog and overlay header to labs in Latte. Updates to dialog/overlay core site examples and guidance and header block pattern page to be updated for core in Lungo sprint.

**Comment by mark-tate** [comment]
Lungo Goal: no dev capacity this sprint

**Comment by bhoppers2008** [comment]
Moved to Macchiato sprint. No capacity in Lungo. @navkaur76 can resume the work when she returns from holiday. @mark-tate

**Comment by mark-tate** [comment]
Macc Goal:

Design/Dev kickoff to re-align on sprint priorities and how we will deliver/QA related tickets and publish in Figma

Plan in place for site updates
Design to review overlay so we can move to Labs
API review by EOS

**Comment by mark-tate** [comment]
Mocca Goal: MTK review from labs

**Comment by yunjungyeh** [comment]
Wed 27 Nov - Scope notes from @bhoppers2008 
- Takes the default bordered appearance (in all cases)
- displayed if action after any button click can’t happen immediately
- after what period should it time out (example(? Check), we make recommendation - recommend use of salt duration tokens-if need check with paul)
- cursor is always arrow (loading button is not clickable)
- loading button doesn’t have states
- what happens if the loading fails? what next...
- ada spec + review with jake (its not clickable, should it receive focus? button had focus beforehand, on enter/space press what happens to focus?)
- docs - see if you can reuse any of ivans examples (probably limited examples... when you click a button what happens)
- figma - check with josh if its a) part of button or b) a separate component
- a) loading button is new component set in button (not variant)
- b) its own component

**Comment by bhoppers2008** [comment]
Need to make changes to the header block status to allow for multiline. Meeting tomorrow (Tuesday) with plan to merge in Figma. Nav to keep Fernanda up to date.

**Comment by mark-tate** [comment]
Red Eye: BLOCKED - resolution for icon sizes to text sizes #4433 will unblock this story

**Comment by mark-tate** [comment]
@Fercas123 baseline alignment - how easy would it be to use this to align text and icons ?

**Comment by mark-tate** [comment]
Red Eye: Proposal to unblock by EOS

**Comment by origami-z** [comment]
`<DialogHeader` search returns 743 instances, need to be careful about breaking change

**Comment by bhoppers2008** [comment]
Header Block merged to labs. 

@joshwooding  New ticket to be created to move to core.
@navkaur76 New ticket to be created for header block updates in new Salt DS library (@yunjungyeh)
